fv3gfs-util/fv3gfs/util/partitioner.py

```diff
@@ -812,17 +968,66 @@ def subtile_slice(
-            to both ranks, or only to the higher rank (default)
+        edge_interior_ratio (optional): ratio between interior and boundary tile sizes.
+            Assumes full integer divisibility for both interior
+            and boundary tile sizes. Default is 1.0, meaning equal subtiles.
+        overlap (optional): if True, for interface variables include the part
+                of the array shared by adjacent ranks in both ranks. If False, ensure
+                only one of those ranks (the greater rank) is assigned the overlapping
+                section. Default is True.
     """
+
     return_list = []
+    subtile_extent = subtile_extents_from_tile_metadata(
+        dims, global_extent, layout, edge_interior_ratio
+    )
     # discard last index for interface variables, unless you're the last rank
     # done so that only one rank is responsible for the shared interface point
-    for index in _index_generator(dims, global_extent, subtile_index, layout):
-        start = index.i_subtile * index.base_extent
-        if index.is_end_index or overlap:
-            end = start + index.extent
+    num_decomposed_dims = int(len(subtile_extent) / 2)
+    horizontal_dim_index = 0
+
+    for num_dim, (dim, dim_extent) in enumerate(zip(dims, global_extent)):
+        if dim in constants.HORIZONTAL_DIMS:
+            if (
+                subtile_index[horizontal_dim_index] == 0
+                or layout[horizontal_dim_index] < 3
+            ):
+                # this is technically not the edge tile size, but does not matter as subtile_index for that dim is 0.
+                # alternatively equal sized tiles go here, which can take both subdomain sizes.
+                start = subtile_index[horizontal_dim_index] * subtile_extent[num_dim]
+            else:
+                start = (subtile_index[horizontal_dim_index] - 1) * subtile_extent[
+                    num_dim
+                ] + subtile_extent[num_dim + num_decomposed_dims]
+
+            is_end_index = subtile_index[horizontal_dim_index] == (
+                layout[horizontal_dim_index] - 1
+            )
+            if layout[horizontal_dim_index] < 3 or (
+                subtile_index[horizontal_dim_index] == 0 or is_end_index
+            ):
+                end = start + _interface_overlap_extent(
+                    dim,
+                    is_end_index,
+                    subtile_extent[num_dim + num_decomposed_dims],
+                    overlap,
+                )
+            else:
+                end = start + _interface_overlap_extent(
+                    dim, False, subtile_extent[num_dim], overlap
+                )
+            horizontal_dim_index = horizontal_dim_index + 1
         else:
-            end = start + index.base_extent
-        return_list.append(slice(start, end))
+            start = 0
+            end = dim_extent
+
+        return_list.append(slice(int(start), int(end)))
```

Comment:
This code looks very similar to the other large, complex block of code. Are there any shared functionalities you can refactor out into a helper routine? It's hard for me to tell because I can't quite tell what the codes are doing.

Reply:
I've added some more in-code comments to this block of code, hopefully it's more clear. The two blocks of code do a similar thing, but not exactly. If needed I can probably refactor it slightly into a helper routine.